interfaces/builtin/shared_memory.go

```diff
@@ -0,0 +1,161 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+const sharedMemorySummary = `allows two snaps to use predefined shared memory objects`
+
+const sharedMemoryBaseDeclarationSlots = `
+  shared-memory:
+    allow-installation:
```

Comment:
hmm need to think about this, what we typically do for super-privileged snaps is to use `allow-installation: false`, but usually that's on the _plug_ side not on the _slot_ side, but here I think we want the super-privilege settings on the slot side. 

This does have some precedence, we do this for example with the `lxd` and `docker` interfaces, so I think what you want for the base-declaration for the slot side is:

```yaml
  shared-memory:
    allow-installation: false
    deny-connection: true
    deny-auto-connection: true
```

but that does leave a question about what to use for setting up auto-connection rules, since the precedence is that the plug over-rules the slot, so if we put that into the slot, then we could try something like this in the plug declaration:

```yaml
  shared-memory:
    allow-installation: true
    allow-connection:
      plug-attributes:
        shared-memory: $SLOT(shared-memory)
    allow-auto-connection: 
      plug-publisher-id:
        - $SLOT_PUBLISHER_ID
      plug-attributes:
        content: $SLOT(content)
```

but importantly because the plug rules will over-ride the slot rules this means that a snap with a plug would immediately over-ride the super-privileged nature of the slot and it would connect. However, I think it still makes sense because the `allow-installation: false` on the slot means that the snap with the slot can not even be installed without a store snap-declaration assertion overriding the `allow-installation` setting to `true`, so maybe this all makes sense.

In any case I would like to hear from @pedronis to confirm my understanding is correct and that the two snippets above are what we should put for the slot and plug side respectively.

Reply:
yes, those rules would make a snap with a plug follow rules like the content interface,  notice that I think you didn't fully update the snippet, we want I think:

```
allow-auto-connection: 
      plug-publisher-id:
        - $SLOT_PUBLISHER_ID
      plug-attributes:
        share-memory: $SLOT(shared-memory)
```

slot snaps cannot be installed by default, as they are the ones that can select the shared file, it seems ok

Reply:
Done!